### PR TITLE
`poolswap` `pool.add.liquidity` `pool.remove.liquidity`

### PR DESCRIPTION
**Pull request by monstrobishi**
#### What kind of PR is this?:

/kind feature

#### What this PR does / why we need it:

Adds indexing for poolswap, pool.add.liquidity and pool.remove.liquidity

#### Which issue(s) does this PR fixes?:

Fixes #

### CODE REVIEW

**Comment by fuxingloh** [comment]
We have to wait for a FC Node release to implement this, for it to work with the latest loans changes.

**Comment by monstrobishi** [comment]
@fuxingloh could you elaborate please? I don't think this has any impact on loans changes

**Comment by fuxingloh** [comment]
> @fuxingloh could you elaborate please? I don't think this has any impact on loans changes

When FC release:

1. the pool balance will not tally if we don't track composites swap
2. there is a loan mechanism that uses swap to burn tokens

File: src/module.indexer/model/dftx/pool.add.liquidity.ts [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        continue
+      }
+
+      const forward = poolPair.tokenA.id === token1.balances[0].token
+      const tokenA = forward ? token1 : token2
+      const tokenB = forward ? token2 : token1
+
+      let liquidity = new BigNumber(0)
+      let totalLiquidity = new BigNumber(poolPair.totalLiquidity)
+      if (totalLiquidity.isEqualTo(0)) {
```
Comment (ivan-zynesis) [inline src/module.indexer/model/dftx/pool.add.liquidity.ts:45]: Wonder how we keep  all these stick with consensus updates if any.

File: src/module.indexer/model/dftx/pool.add.liquidity.ts [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+    let liquidity = new BigNumber(0)
+    let totalLiquidity = new BigNumber(poolPair.totalLiquidity)
+    if (totalLiquidity.isEqualTo(0)) {
+      liquidity = tokenA.balances[0].amount.times(tokenB.balances[0].amount).sqrt()
+      liquidity = liquidity.minus(MINIMUM_LIQUIDITY)
+      totalLiquidity = new BigNumber(MINIMUM_LIQUIDITY)
+    } else {
+      const liqA = tokenA.balances[0].amount.times(totalLiquidity.dividedBy(poolPair.tokenA.reserve))
+      const liqB = tokenB.balances[0].amount.times(totalLiquidity.dividedBy(poolPair.tokenA.reserve))
```
Comment (canonbrother) [inline src/module.indexer/model/dftx/pool.add.liquidity.ts:52]: should it divided by tokenB.reserve??

```suggestion
      const liqB = tokenB.balances[0].amount.times(totalLiquidity.dividedBy(poolPair.tokenB.reserve))
```

File: src/module.indexer/model/dftx/pool.add.liquidity.ts [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+    const forward = poolPair.tokenA.id === token1.balances[0].token
+    const tokenA = forward ? token1 : token2
+    const tokenB = forward ? token2 : token1
+
+    let liquidity = new BigNumber(0)
+    let totalLiquidity = new BigNumber(poolPair.totalLiquidity)
+    if (totalLiquidity.isEqualTo(0)) {
+      liquidity = tokenA.balances[0].amount.times(tokenB.balances[0].amount).sqrt()
+      liquidity = liquidity.minus(MINIMUM_LIQUIDITY)
```
Comment (canonbrother) [inline src/module.indexer/model/dftx/pool.add.liquidity.ts:48]: i don't understand this :sweat_smile: 
why do we need deduct another 0.00001??

Comment (canonbrother) [inline src/module.indexer/model/dftx/pool.add.liquidity.ts:48]: i write this way.. clearer
```ts
let totalLiquidity = new BigNumber(MINIMUM_LIQUIDITY)
if (new BigNumber(poolPair.totalLiquidity).isEqualTo(0)){
    liquidity = ...
} else {
    liquidity = ...
}
totalLiquidity = totalLiquidity.plus(liquidity)
```

File: src/module.indexer/model/dftx/update.poolpair.spec.ts [hunk 4]
```diff
@@ -76,51 +85,69 @@ describe('invalidate', () => {
...
       block: expect.any(Object)
     })
 
-    await invalidateFromHeight(app, container, heightUpdated - 1)
+    await invalidateFromHeight(app, container, heightUpdated)
     await container.generate(2)
     await waitForIndexedHeight(app, heightUpdated)
 
-    const poolPairInvalidated = await poolPairMapper.getLatest('11')
+    const poolPairInvalidated = await poolPairMapper.getLatest('8')
```
Comment (canonbrother) [inline src/module.indexer/model/dftx/update.poolpair.spec.ts:127]: i see.. good reference for me

File: src/module.indexer/model/dftx/poolpair.e2e.ts [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+      sort: '00000003'
+    })
+
+    const removeMsg = await removePoolLiquidity(container, {
+      address: ownerAddress,
+      tokenLP: result?.poolPairId ?? '',
+      amountLP: new BigNumber(20)
+    })
+
+    console.log(removeMsg)
```
Comment (canonbrother) [inline src/module.indexer/model/dftx/poolpair.e2e.ts:211]:
```suggestion
```

**Pull request closed without merging**